File: src/main/scala/com/github/mauricio/postgresql/column/DateEncoderDecoder.scala
package com.github.mauricio.postgresql.column

import org.joda.time.{ReadableInstant, LocalDate}
import org.joda.time.format.DateTimeFormat
import com.github.mauricio.postgresql.exceptions.DateEncoderNotAvailableException

/**
 * User: Maurício Linhares
 * Date: 3/5/12
 * Time: 6:12 PM
 */

object DateEncoderDecoder extends ColumnEncoderDecoder {

  private val formatter = DateTimeFormat.forPattern("yyyy-MM-dd")

  override def decode(value: String): LocalDate = {
    this.formatter.parseLocalDate(value)
  }

  override def encode( value : Any ) : String = {
    value match {
      case d : java.sql.Date => this.formatter.print( new LocalDate(d) )
      case d : ReadableInstant => this.formatter.print(d)
      case _ => throw new DateEncoderNotAvailableException(value)
    }
  }

}

File: db-async-common/src/main/java/com/github/jasync/sql/db/column/DateEncoderDecoder.kt
package com.github.jasync.sql.db.column

import com.github.jasync.sql.db.exceptions.DateEncoderNotAvailableException
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAccessor

object DateEncoderDecoder : ColumnEncoderDecoder {

    private const val ZeroedDate = "0000-00-00"

    private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd") // DateTimeFormat.forPattern("yyyy-MM-dd")

    override fun decode(value: String): LocalDate? =
        if (ZeroedDate == value) {
            null
        } else {
            LocalDate.parse(value, this.formatter)
        }

    override fun encode(value: Any): String {
        return when (value) {
            is java.sql.Date -> value.toLocalDate().format(this.formatter)
            is TemporalAccessor -> this.formatter.format(value)
            else -> throw DateEncoderNotAvailableException(value)
        }
    }

}
